### PR TITLE
A Fistful of Shotgun Shells - Nerva Edition

### PR DESCRIPTION
**Pull request by Cheedows**
Basically what it says on the tin, you can now stack shotgun shells/slugs together into a bundle of up to 4 shells. All types of shotgun shells can be stacked in this bundle, EMPs, blanks, buckshot, stun shells you name it. As part of this you get a very aesthetically pleasing dynamic sprite that will correctly render every different type of shell you have in your bundle. All credits for the lovely sprites goes to Dongwaiver.

These bundles of shells are dynamic, their sprite will change and is rendered based off of what shells you have stacked (for example, 3 slugs and 1 buckshot will be shown as 3 slugs and 1 buckshot in the item handful). Part of this is also that while they are sort of like speedloaders (in this case, shotloaders), there is no speedloader item itself. When you have used up all 4 shells or split them apart manually the shot bundle object disappears, and you are left with just shells. It doesn't leave a mess and pretty seamless to use.

Part of this update also slightly changes how shotguns load in ammo. For just the case of shotguns, you can load in either single shells one at a time like previous, or apply the handful of shells directly to the shotgun and it will load it one by one. A side effect of this is that shotholders will also be able to do this (previously shotholders could only reload double barrels, as it's considered a speedloader. Now it can also reload combat shotguns like the new handfuls, albeit only one shell at a time). Overall, the aim of this is to make shotguns more convenient to use without the hassle of managing a bunch of individual shells, and from anyone who has played Colmar, it works similar to the handful of slugs/casings that they have.

The downside of this is that these new handfuls will work similarly to the existing shotholders we have, which might make the shot holders obsolete. However, handfuls are slightly larger (normal sized instead of small like shotholders), which makes more sense, it's just a bundle of shells rather than a neatly packed array into a holder. The side effect of this balance wise is also that handfuls only fit in your backpack, while shot holders can still fit in your pocket like a magazine.

### CODE REVIEW

File: code/modules/projectiles/guns/projectile/shotgun.dm [hunk 1]
```diff
@@ -2,6 +2,43 @@
...
+		if(loaded.len >= max_shells)
+			to_chat(user, "<span class='warning'>[src] is full.</span>")
+			return
+		if(!user.unEquip(C, src))
+			return
+		loaded.Insert(1, C) //add to the head of the list
+		user.visible_message("[user] inserts \a [C] into [src].", "<span class='notice'>You insert \a [C] into [src].</span>")
+		playsound(loc, load_sound, 50, 1)
+	update_icon()
+
```
Comment (Naxodile) [inline code/modules/projectiles/guns/projectile/shotgun.dm:41]:
```suggestion
/obj/item/weapon/gun/projectile/shotgun/load_ammo(var/obj/item/A, mob/user)
	if(istype(A, /obj/item/ammo_magazine/shotbundle))
		. = TRUE
		var/obj/item/ammo_magazine/shotbundle/SB = A
		if(caliber != SB.caliber)
			return

		if(loaded.len >= max_shells)
			to_chat(user, "<span class='warning'>[src] is full!</span>")
			return

		var/obj/item/ammo_casing/C = SB.stored_ammo[SB.stored_ammo.len]
		SB.stored_ammo -= C
		loaded.Insert(1, C)
		user.visible_message("[user] inserts \a [C] into [src].", "<span class='notice'>You insert \a [C] into [src].</span>")
		playsound(loc, load_sound, 50, 1)
		SB.update_icon()
		SB.check_ammo_count()
		return

	return ..()
```

**Review by Naxodile — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/projectile/shotgun.dm (hunk 1 above)
Comment (Naxodile) [reply code/modules/projectiles/guns/projectile/shotgun.dm:41]: This removes the one-by-one for shell holders. You should honestly just mark all shotguns as accepting speedloaders and overwrite the attack_self behavior for shotguns to set allow_dump=0 save for the double barrel.

File: code/modules/projectiles/ammunition.dm [hunk 2]
```diff
@@ -143,9 +143,13 @@
 		if(stored_ammo.len >= max_ammo)
 			to_chat(user, "<span class='warning'>[src] is full!</span>")
 			return
+		if(!C.BB)
+			to_chat(user, "<span class='warning'>[C] is spent.</span>")
+			return
 		if(!user.unEquip(C, src))
 			return
 		stored_ammo.Add(C)
+		user.visible_message("[user] insert [C] to [src].", "<span class='notice'>You insert [C] into [src].</span>")
```
[the remark this answers is not in the document]
Comment (Cheedows) [reply code/modules/projectiles/ammunition.dm]: Removed the message for this to cut down on spam concern.

File: code/modules/projectiles/ammunition/boxes.dm [hunk 3]
```diff
@@ -68,6 +68,64 @@
 	icon_state = "44_r"
 	ammo_type = /obj/item/ammo_casing/c44/rubber
 
+/obj/item/ammo_magazine/shotbundle
+	name = "shotgun shell bundle"
+	desc = "A bundle of 12 gauge shells."
+	icon_state = "slshell"
+	caliber = "shotgun"
+	ammo_type = /obj/item/ammo_casing/shotgun
+	matter = list(MATERIAL_STEEL = 1440)
```
[the remark this answers is not in the document]
Comment (Cheedows) [reply code/modules/projectiles/ammunition/boxes.dm]: Changed it to be 720, which is the equivalent of 2 shells (a bundle will always have at least 2) so net neutral metal gain.

File: code/modules/projectiles/ammunition/boxes.dm [hunk 4]
```diff
@@ -68,6 +68,64 @@
...
+		C.set_dir(pick(GLOB.alldirs))
+	stored_ammo.Cut()
+	if(!user.unEquip(src, src))
+		return
+	qdel(src)
+
+/obj/item/ammo_magazine/shotbundle/on_update_icon()
+	..()
+	overlays.Cut()
+	src.icon_state = "netshell"
```
[the remark this answers is not in the document]
Comment (Cheedows) [reply code/modules/projectiles/ammunition/boxes.dm]: Removed src references from overlays and icon_state

File: code/modules/projectiles/ammunition/boxes.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (Naxodile) [reply code/modules/projectiles/ammunition/boxes.dm]: Not all shells are the same amount, use 2 blanks and you're getting metal gain again. You're just going to need to either have it be a matter of 0 or recalc on additions/removals.

File: code/modules/projectiles/ammunition/boxes.dm [hunk 5]
```diff
@@ -68,6 +68,63 @@
...
+	stored_ammo.Cut()
+	if(!user.unEquip(src, src))
+		return
+	qdel(src)
+
+/obj/item/ammo_magazine/shotbundle/on_update_icon()
+	overlays.Cut()
+	icon_state = "netshell"
+	var/count = 1
+	for(var/obj/item/ammo_casing/C in src.stored_ammo)
```
Comment (Naxodile) [inline code/modules/projectiles/ammunition/boxes.dm:123]:
```suggestion
	for(var/obj/item/ammo_casing/C in stored_ammo)
```

File: code/modules/projectiles/ammunition/boxes.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (Cheedows) [reply code/modules/projectiles/ammunition/boxes.dm]: Made the material 0, confirmed autolathes don't take it.

File: code/modules/projectiles/ammunition/boxes.dm [hunk 6]
```diff
@@ -68,6 +68,64 @@
...
+	icon_state = "slshell"
+	caliber = "shotgun"
+	ammo_type = /obj/item/ammo_casing/shotgun
+	matter = list(MATERIAL_STEEL = 1440)
+	initial_ammo = 0
+	max_ammo = 4
+	multiple_sprites = 1
+	w_class = ITEM_SIZE_NORMAL
+
+/obj/item/ammo_magazine/shotbundle/attack_hand(mob/user)
```
[the remark this answers is not in the document]
Comment (Cheedows) [reply code/modules/projectiles/ammunition/boxes.dm:83]: Implemented the check_ammo_count function and the mob/user param.

File: code/modules/projectiles/ammunition/boxes.dm [hunk 7]
```diff
@@ -68,6 +68,64 @@
...
+			var/obj/item/ammo_casing/C = stored_ammo[stored_ammo.len]
+			stored_ammo-=C
+			user.put_in_hands(C)
+			user.visible_message("\The [user] removes \a [C] from [src].", "<span class='notice'>You remove \a [C] from [src].</span>")
+			update_icon()
+	else
+		..()
+		return
+
+/obj/item/ammo_magazine/shotbundle/attack_self(mob/user)
```
[the remark this answers is not in the document]
Comment (Cheedows) [reply code/modules/projectiles/ammunition/boxes.dm:96]: Same as above, implemented this change to the code.

File: code/modules/projectiles/ammunition/boxes.dm [hunk 8]
```diff
@@ -68,6 +68,44 @@
...
+        qdel(src)
+
+
+/obj/item/ammo_magazine/shotbundle/attack_self(mob/user)
+	..()
+	qdel(src)
+
+/obj/item/ammo_magazine/shotbundle/on_update_icon()
+	overlays.Cut()
+	icon_state = "netshell"
```
Comment (Naxodile) [inline code/modules/projectiles/ammunition/boxes.dm:102]: How did I miss this

Comment (Cheedows) [reply code/modules/projectiles/ammunition/boxes.dm]: Removed setting the dummy icon_state when updating the icon with the overlay.

**Comment by Glloyd** [comment]
Haven't taken a look through the code yet, but move the icons for the shotbundle to a Urist .dmi please.

**Review by Naxodile — CHANGES_REQUESTED** [review]
Invoking a squash request on this PR, all the commits are just code quality

**Comment by Glloyd** [comment]
we can also squash it when we merge it if need be

**Review by Naxodile — APPROVED** [review]

**Comment by Cheedows** [comment]
Oh god oh fuck, can this be reverted real quick @Glloyd @Naxodile . I was paranoid and double checked this morning with a compile, and realized that the compiler complains if you do the icon in double quotes like icon = "icons/urist/items/shotbundle.dmi". It just needs to be switched to single quotes like icon = 'icons/urist/items/shotbundle.dmi', it was from when I moved the icons to the urist dmi and switched the icon file to point to the new one.

**Pull request merged**